presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/SetOperationNodeTranslator.java

```diff
@@ -0,0 +1,176 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.metadata.Signature;
+import io.prestosql.spi.type.StandardTypes;
+import io.prestosql.spi.type.Type;
+import io.prestosql.sql.planner.PlanNodeIdAllocator;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.SymbolAllocator;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.SetOperationNode;
+import io.prestosql.sql.planner.plan.UnionNode;
+import io.prestosql.sql.tree.Cast;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.FunctionCall;
+import io.prestosql.sql.tree.NullLiteral;
+import io.prestosql.sql.tree.QualifiedName;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.concat;
+import static io.prestosql.metadata.FunctionKind.AGGREGATE;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+import static io.prestosql.spi.type.TypeSignature.parseTypeSignature;
+import static io.prestosql.sql.planner.plan.AggregationNode.singleGroupingSet;
+import static io.prestosql.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static java.util.Objects.requireNonNull;
+
+public class SetOperationNodeTranslator
+{
+    private static final String MARKER = "marker";
+    private static final Signature COUNT_AGGREGATION = new Signature("count", AGGREGATE, parseTypeSignature(StandardTypes.BIGINT), parseTypeSignature(StandardTypes.BOOLEAN));
+
+    private final SymbolAllocator symbolAllocator;
+    private final PlanNodeIdAllocator idAllocator;
+
+    public SetOperationNodeTranslator(SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        this.symbolAllocator = requireNonNull(symbolAllocator, "SymbolAllocator is null");
+        this.idAllocator = requireNonNull(idAllocator, "PlanNodeIdAllocator is null");
+    }
+
+    public TranslatorResult simplify(SetOperationNode node)
+    {
+        checkArgument(!(node instanceof UnionNode), "Cannot simplify a UnionNode");
```

Comment:
At some point, we should change the meaning of `UnionNode`. Currently, it represents the concatenation of two tables, but it should really mean the "set union" of two tables. There should be another operation type to represent concatenation.

Then, this logic would just work for all set operations.